### PR TITLE
First pass at implementing the /clone command

### PR DESCRIPTION
**Pull request by smartboyathome**
So, this is a big one that requires a lot of manual testing (tried writing some unit tests, but the mocks just weren't that easy to create). This implementation shouldn't require too much memory when executing, due to the dynamic direction of the iteration in order to avoid any overlap. By making it dynamic for each axis, I was able to clone without staging copies of each block in memory first. As mentioned, I did as much manual testing as I could, but doubtless I have missed several scenarios that may break this. More eyes are needed on this change before it should be merged.

### CODE REVIEW

File: src/main/java/net/glowstone/command/minecraft/CloneCommand.java [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+        public BackwardsAxisIterator(int length) {
+            this.current = length;
+        }
+
+        @Override
+        protected Integer computeNext() {
+            if (current >= 0) {
+                // Could use a post-decrement operator here, but this is a bit more clear in getting the meaning across.
+                int retval = current;
+                current -= 1;
```
Comment (aramperes) [inline src/main/java/net/glowstone/command/minecraft/CloneCommand.java]: I understand why you'd prefer not using `int retval = current--;`, but maybe replace the `current -= 1` line with `current--;`

File: src/main/java/net/glowstone/command/minecraft/CloneCommand.java [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+            return false;
+        }
+
+        if (args.length < 9) {
+            sender.sendMessage(ChatColor.RED + "Usage: " + usageMessage);
+            return false;
+        }
+
+        if (!CommandUtils.isPhysical(sender)) {
+            sender.sendMessage("Cannot access blocks outside of the world");
```
Comment (aramperes) [inline src/main/java/net/glowstone/command/minecraft/CloneCommand.java]: I think the error message and the if-statement don't match here. (e.g. should be `This command may only be executed by physical objects` or something. I think other commands have this message somewhere.

File: src/main/java/net/glowstone/command/minecraft/CloneCommand.java [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+                if (args.length >= 12) {
+                    Material blockType = ItemIds.getItem(args[11]);
+                    if (args.length >= 13) {
+                        Byte data = null;
+                        try {
+                            data = Byte.parseByte(args[12]);
+                        } catch (NumberFormatException ignored) {
+                        }
+                        if (data == null || 0 > data || data > 15) {
+                            sender.sendMessage(ChatColor.RED + "Usage: " + usageMessage);
```
Comment (aramperes) [inline src/main/java/net/glowstone/command/minecraft/CloneCommand.java]: Give a more precise error here (e.g. `Invalid data: <data>`)

File: src/main/java/net/glowstone/command/minecraft/CloneCommand.java [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+                        blockFilter = new FilteredBlockFilter(blockType);
+                    }
+                } else {
+                    sender.sendMessage(ChatColor.RED + "Usage: " + usageMessage);
+                    return false;
+                }
+                break;
+
+            default:
+                sender.sendMessage(ChatColor.RED + "Usage: " + usageMessage);
```
Comment (aramperes) [inline src/main/java/net/glowstone/command/minecraft/CloneCommand.java:167]: Also here, say something like `Unknown block filter: <filter>`

File: src/main/java/net/glowstone/command/minecraft/CloneCommand.java (hunk 1 above)
Comment (smartboyathome) [reply src/main/java/net/glowstone/command/minecraft/CloneCommand.java]: Yeah, my biggest problem with the post- and pre- increment/decrement operators is that, in variable assignments, its not obvious which returns the old value and which returns the new value. I can definitely change that line to be current-- though.

File: src/main/java/net/glowstone/command/minecraft/CloneCommand.java [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+        }
+    }
+
+    public CloneCommand() {
+        super(
+                "clone",
+                "Clones a section of the world.", "/clone <x1> <y1> <z1>  <x2> <y2> <z2>  <x> <y> <z> [maskMode] [cloneMode] [tileName] [tileData]",
+                Collections.emptyList()
+        );
+        setPermission("minecraft.command.seed");
```
Comment (aramperes) [inline src/main/java/net/glowstone/command/minecraft/CloneCommand.java]: Wrong permission :^)

File: src/main/java/net/glowstone/command/minecraft/CloneCommand.java [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+                            fromBlock.setType(Material.AIR, false);
+                        }
+
+                        blocksCloned++;
+                    }
+                }
+            }
+        }
+
+        sender.sendMessage("Cloned " + blocksCloned + " blocks.");
```
Comment (aramperes) [inline src/main/java/net/glowstone/command/minecraft/CloneCommand.java]: If `blocksCloned` ends up being 0, should another message pop up here? (e.g. No blocks were cloned)

**Review by mastercoms — APPROVED** [review]

**Pull request merged**